### PR TITLE
Various implementations

### PR DESCRIPTION
**Pull request by Grayahm**
- Roseanne's Research
- Night Maintenance
- Bebe's Search

### CODE REVIEW

File: src/tcgwars/logic/impl/gen4/SecretWonders.groovy [hunk 1]
```diff
@@ -2679,17 +2679,26 @@ public enum SecretWonders implements LogicCardInfo {
         return basicTrainer (this) {
           text "You can play only one Supporter card each turn. When you play this card, put it next to your Active Pokémon. When your turn ends, discard this card.\nChoose a card from your hand and put it on top of your deck. Search your deck for a Pokémon, show it to your opponent, and put it into your hand. Shuffle your deck afterward. (If this is the only card in your hand, you can’t play this card.)"
           onPlay {
+            my.hand.getExcludedList(thisCard).select("Choose the card to put back in your deck").showToOpponent("Chosen card")
```
Comment (luongthomas) [inline src/tcgwars/logic/impl/gen4/SecretWonders.groovy:2682]: I believe you need to add `.moveTo(my.deck)` at the end here

File: src/tcgwars/logic/impl/gen4/SecretWonders.groovy [hunk 2]
```diff
@@ -2679,17 +2679,26 @@ public enum SecretWonders implements LogicCardInfo {
...
+            my.deck.search(count:1,"Choose the Pokémon to put back in your deck",cardTypeFilter(POKEMON)).showToOpponent("Chosen card")
           }
           playRequirement{
+            assert my.hand.size() > 1 : "You can't play this card"
```
Comment (luongthomas) [inline src/tcgwars/logic/impl/gen4/SecretWonders.groovy:2686]: Could you change this message to say `Your hand needs one other card to play this card` or something along those lines?

File: src/tcgwars/logic/impl/gen4/SecretWonders.groovy [hunk 3]
```diff
@@ -2679,17 +2679,26 @@ public enum SecretWonders implements LogicCardInfo {
         return basicTrainer (this) {
           text "You can play only one Supporter card each turn. When you play this card, put it next to your Active Pokémon. When your turn ends, discard this card.\nChoose a card from your hand and put it on top of your deck. Search your deck for a Pokémon, show it to your opponent, and put it into your hand. Shuffle your deck afterward. (If this is the only card in your hand, you can’t play this card.)"
           onPlay {
+            my.hand.getExcludedList(thisCard).select("Choose the card to put back in your deck").showToOpponent("Chosen card")
+            my.deck.search(count:1,"Choose the Pokémon to put back in your deck",cardTypeFilter(POKEMON)).showToOpponent("Chosen card")
```
Comment (luongthomas) [inline src/tcgwars/logic/impl/gen4/SecretWonders.groovy:2683]: This needs `.moveTo(my.hand)` at the end

File: src/tcgwars/logic/impl/gen4/SecretWonders.groovy [hunk 4]
```diff
@@ -2727,8 +2736,11 @@ public enum SecretWonders implements LogicCardInfo {
         return basicTrainer (this) {
           text "You can play only one Supporter card each turn. When you play this card, put it next to your Active Pokémon. When your turn ends, discard this card.\nSearch your deck for up to 2 in any combination of Basic Pokémon and basic Energy cards, show them to your opponent, and put them into your hand. Shuffle your deck afterward."
           onPlay {
+            my.deck.search(max: 2,"Search your deck for up to 2 in any combination of Basic Pokémon and basic Energy cards",{it.cardTypes.is(BASIC_ENERGY) || it.cardTypes.is(BASIC)}).moveTo(hand)
+            shuffleDeck()
           }
           playRequirement{
+            assert my.deck
```
Comment (luongthomas) [inline src/tcgwars/logic/impl/gen4/SecretWonders.groovy:2743]: Add an assert message here like `Your deck is empty.`

File: src/tcgwars/logic/impl/gen4/SecretWonders.groovy (hunk 1 above)
Comment (Grayahm) [reply src/tcgwars/logic/impl/gen4/SecretWonders.groovy:2682]: Fixed

**Pull request merged**